tensorflow_addons/layers/embeddingbag.py

```diff
@@ -0,0 +1,130 @@
+import tensorflow as tf
+from tensorflow.python.keras import constraints
+from tensorflow.python.keras import initializers
+from tensorflow.python.keras import regularizers
+from typeguard import typechecked
+from tensorflow_addons.utils.resource_loader import LazySO
+
+_embeddingbag_so = LazySO("custom_ops/layers/_embeddingbag_ops.so")
+
+
+def embeddingbag(
+    indices,
+    values,
+    weights,
+    name=None,
+):
+    """EmbeddingBag computation
+
+    See [PyTorch op](https://pytorch.org/docs/stable/generated/torch.nn.EmbeddingBag.html).
+
+    Equivalent to tf.gather() followed by tf.reduce_sum() across the last dimension, with optional
+    weights. Fusing these into a single op has massive benefits for execution speed and particularly
+    memory usage, as the intermediate output of the gather never needs to be materialized.
+
+    Args:
+      indices: An int32 or int64 `Tensor` of the indices to gather from
+          `values`. Must be at least 2-dimensional, as the last dimension
+          will be summed out. Maximum value must be less than values.shape[0].
+      values: A float32 `Tensor` from which to gather values. Must be rank 2.
+      weights: A float32 `Tensor` of weights which will be applied to each of
+          the gathered embedding vectors before the sum step.
+      name: A name for the operation (optional).
+
+    Returns:
+      A `Tensor` of the format specified by `data_format`.
+    """
+
+    # TODO Assert weights have same shape as indices, if present
+    with tf.name_scope(name or "embedding_bag"):
+        op_call = _embeddingbag_so.ops.addons_embedding_bag
+
+        if weights is None:
+            weights = tf.ones_like(indices, dtype=tf.float32)
+
+        ret = op_call(
+            indices,
+            values,
+            weights
+        )
+        return ret
+
+
+@tf.RegisterGradient("Addons>EmbeddingBag")
+def _embedding_bag_grad(op, grad):
+    indices, values, weights = op.inputs[:3]
+    op_call = _embeddingbag_so.ops.addons_embedding_bag_grad
+    values_grad, weights_grad = op_call(indices, values, weights, grad)[:2]  # Drop the dummy outputs
+    return [None, values_grad, weights_grad]
+
+
+@tf.keras.utils.register_keras_serializable(package="Addons")
+class EmbeddingBag(tf.keras.layers.Layer):
+    """EmbeddingBag Layer
+
+    See [PyTorch op](https://pytorch.org/docs/stable/generated/torch.nn.EmbeddingBag.html).
+
+    Equivalent to tf.gather() followed by tf.reduce_sum() across the last dimension, with optional
+    weights. Fusing these into a single op has massive benefits for execution speed and particularly
+    memory usage, as the intermediate output of the gather never needs to be materialized.
+
+    Input Shapes:
+      indices: An int32 or int64 `Tensor` of the indices to gather from
+          `values`. Must be at least 2-dimensional, as the last dimension
+          will be summed out. Maximum value must be less than values.shape[0].
```

Comment:
If I understand correctly, there are only two cases in PyTorch:

1.  `(batch_size, sequence_length)`
2. `(batch_size,)`:  in this case, it has same effects as `(batch_size, 1)`

Any reason or use case that you want to  support higher dimension indices?

Reply:
You're correct that I'm not precisely following the PyTorch semantics. The reason was to more closely follow the semantics of tf.gather, where the indices tensor can be N-dimensional. My op treats the final dimension as the 'bag' dimension, which is analogous to PyTorch when N = 2, and any other dimensions are all treated as 'batch' dimensions. The C++/CUDA code is completely blind to these details - it only cares about the total size of the indices array and the length of the final dimension, and treats all other dimensions as one single 'batch' dimension.

If people prefer, I can try to mimic the PyTorch semantics more closely, but it would probably take some hacking to support a separate offset tensor.

Reply:
I know that your impl is independent to the first (N-1)-dim, but note that if you want to support higher dimension, there will be less compiler optimization kickin for Eigen compilation (If you don't specialize each rank's functor).

For 1-D indices, my thought is that we can implement later :-)

Reply:
I'm happy to constrain the implementation to 2D if you think there's a significant benefit - users can just flatten the first N-1 dimensions and unflatten them afterwards if they have multidimensional input.

Reply:
Sure, I will modify interface later.